File: schemas/plan-override.ts
/**
 * Firestore path: users/{uid}/planOverrides/{planId}
 * Mirror: ios/Shared/PlanOverride.swift
 *
 * Rescheduled plan workouts. Identity per entry is (week, day) — all workouts
 * scheduled that day shift together. Keyed by planId because a user may cycle
 * through multiple library plans and each has its own override sheet.
 */
export interface PlanOverrideDoc {
  entries: PlanOverrideEntry[];
  updatedAt?: string; // ISO date
}

export interface PlanOverrideEntry {
  week: number;         // 1..durationWeeks within the plan
  day: 'Mon' | 'Tue' | 'Wed' | 'Thu' | 'Fri' | 'Sat' | 'Sun';
  shiftedTo: string;    // ISO YYYY-MM-DD — the date the session now occurs on
  shiftedAt: string;    // ISO timestamp — when the shift was recorded
}

File: ios/Shared/PlanOverride.swift
import Foundation

/// Firestore path: users/{uid}/planOverrides/{planId}
/// Mirror: schemas/plan-override.ts
struct PlanOverrideDoc: Equatable {
    var entries: [PlanOverrideEntry]
    var updatedAt: String?

    init(from data: [String: Any]) {
        let rawEntries = data["entries"] as? [[String: Any]] ?? []
        self.entries = rawEntries.compactMap(PlanOverrideEntry.init(from:))
        self.updatedAt = data["updatedAt"] as? String
    }

    init(entries: [PlanOverrideEntry], updatedAt: String? = nil) {
        self.entries = entries
        self.updatedAt = updatedAt
    }

    func toFirestoreData() -> [String: Any] {
        var out: [String: Any] = [
            "entries": entries.map { $0.toFirestoreData() },
        ]
        if let updatedAt { out["updatedAt"] = updatedAt }
        return out
    }
}

struct PlanOverrideEntry: Equatable {
    var week: Int
    var day: String     // "Mon" | "Tue" | ... | "Sun"
    var shiftedTo: String   // YYYY-MM-DD
    var shiftedAt: String   // ISO timestamp

    init(week: Int, day: String, shiftedTo: String, shiftedAt: String) {
        self.week = week
        self.day = day
        self.shiftedTo = shiftedTo
        self.shiftedAt = shiftedAt
    }

    init?(from data: [String: Any]) {
        guard let week = data["week"] as? Int,
              let day = data["day"] as? String,
              let shiftedTo = data["shiftedTo"] as? String,
              let shiftedAt = data["shiftedAt"] as? String else { return nil }
        self.week = week
        self.day = day
        self.shiftedTo = shiftedTo
        self.shiftedAt = shiftedAt
    }

    func toFirestoreData() -> [String: Any] {
        [
            "week": week,
            "day": day,
            "shiftedTo": shiftedTo,
            "shiftedAt": shiftedAt,
        ]
    }
}
